### PR TITLE
[CP-1837] Fix MTP crashes

### PR DESCRIPTION
**Pull request by lukasmas**
Fix the problem with database entries multiplication in mtp_db.
New entries were created without checking if they already existed in the database.

### CODE REVIEW

**Review by awulkiew — APPROVED** [review]

File: mtp/mtp_db.cpp [hunk 1]
```diff
@@ -1,57 +1,79 @@
...
+
     std::optional<std::filesystem::path> FileDatabase::get_filename(Handle handle) const
     {
-        if (const auto result = entries.find(DatabaseEntry{handle}); result != entries.end()) {
-            return result->get_filename();
+        const auto handleToFilenameIter = handleToFilename.find(handle);
+        if (handleToFilenameIter != handleToFilename.end()) {
+            return handle_to_filename::getFilename(handleToFilenameIter);
         }
         return std::nullopt;
```
Comment (marziel) [inline mtp/mtp_db.cpp:42]: [minor]
Probably lines 39-42 can be a single line using the ternary operator `? :`

File: mtp/mtp_db.cpp [hunk 2]
```diff
@@ -1,57 +1,79 @@
...
+    {
+        inline Handle &getHandle(std::map<std::filesystem::path, Handle>::iterator iter)
+        {
+            return iter->second;
+        }
+    } // namespace filename_to_handle
+
+    namespace handle_to_filename
+    {
+        inline auto &getIter(std::map<Handle, std::map<std::filesystem::path, Handle>::iterator>::iterator iter)
```
Comment (marziel) [inline mtp/mtp_db.cpp:19]: This `std::map<Handle, std::map<std::filesystem::path, Handle>::iterator>::iterator` is used in a few places and is quite long - please define an alias with `using`

File: mtp/mtp_db.cpp [hunk 3]
```diff
@@ -1,57 +1,79 @@
...
+        if (handleToFilenameIter != handleToFilename.end()) {
+            return handle_to_filename::getFilename(handleToFilenameIter);
         }
         return std::nullopt;
     }
     bool FileDatabase::remove(const Handle handle)
     {
-        return entries.erase(DatabaseEntry{handle}) != 0;
+        const auto handleToFilenameIter = handleToFilename.find(handle);
+        if (handleToFilenameIter != handleToFilename.end()) {
```
Comment (marziel) [inline mtp/mtp_db.cpp:47]: [minor]
If you negated the condition and returned false, a smaller portion of code would be indented - better readability

File: mtp/mtp_db.cpp [hunk 4]
```diff
@@ -1,57 +1,79 @@
...
-            remove(result->get_handle());
-            entries.insert(DatabaseEntry{result->get_handle(), filename});
+        const auto handleToFilenameIter = handleToFilename.find(handle);
+        if (handleToFilenameIter != handleToFilename.end()) {
+            filenameToHandle.erase(handle_to_filename::getIter(handleToFilenameIter));
+            auto entry = filenameToHandle.emplace(filename, handle);
+            if (!entry.second) {
+                filename_to_handle::getHandle(entry.first) = handle_to_filename::getHandle(handleToFilenameIter);
+            }
+            handle_to_filename::getIter(handleToFilenameIter) = entry.first;
```
Comment (marziel) [inline mtp/mtp_db.cpp:74]: In the previous implementation we had a call to `remove()`, why is none such here?

Comment (lukasmas) [reply mtp/mtp_db.cpp:74]: In the previous implementation, function remove() was simply doing erase(), right now as it has two maps to keep the data we are doing erase in place as function remove() is doing more, and there is no need to remove all records and add them once more within update operation.

File: mtp/mtp_db.cpp (hunk 1 above)
Comment (lukasmas) [reply mtp/mtp_db.cpp:42]: It is also 3 lines after formatting, and personally, I find it more readable as it is now :) 
`        return handleToFilenameIter != handleToFilename.end()
                   ? std::optional<std::filesystem::path>{handle_to_filename::getFilename(handleToFilenameIter)}
                   : std::nullopt;`

File: mtp/mtp_db.cpp (hunk 4 above)
Comment (marziel) [reply mtp/mtp_db.cpp:74]: Fine

File: mtp/mtp_db.cpp (hunk 1 above)
Comment (marziel) [reply mtp/mtp_db.cpp:42]: OK :)

File: mtp/mtp_db.cpp (hunk 2 above)
Comment (lukasmas) [reply mtp/mtp_db.cpp]: Done

File: mtp/mtp_db.cpp (hunk 3 above)
Comment (lukasmas) [reply mtp/mtp_db.cpp]: Done

**Review by marziel — APPROVED** [review]

**Pull request merged**